src/modules/zfs/ZfsJob.cpp

```diff
@@ -0,0 +1,326 @@
+
+    if ( process.exitStatus() == QProcess::CrashExit )
+    {
+        return { false, tr( "The output from the crash was: " ) + output };
+    }
+
+    auto exitcode = process.exitCode();
+    if ( exitcode != 0 )
+    {
+        cWarning() << "Failed to run zpool create.  The output was: " + output;
+        return { false, tr( "Failed to create zpool on " ) + deviceName };
+    }
+
+    return { true, QString() };
+}
+
+Calamares::JobResult
+ZfsJob::exec()
+{
+    QVariantList partitions;
+    Calamares::GlobalStorage* gs = Calamares::JobQueue::instance()->globalStorage();
+    if ( gs && gs->contains( "partitions" ) && gs->value( "partitions" ).canConvert( QVariant::List ) )
+    {
+        partitions = gs->value( "partitions" ).toList();
+    }
+    else
+    {
+        cWarning() << "No *partitions* defined.";
+        return Calamares::JobResult::internalError( tr( "Configuration Error" ),
+                                                    tr( "No partitions are available for Zfs." ),
+                                                    Calamares::JobResult::InvalidConfiguration );
+    }
+
+    const CalamaresUtils::System* system = CalamaresUtils::System::instance();
+
+    QVariantList poolNames;
+
+    for ( auto& partition : qAsConst( partitions ) )
+    {
+        QVariantMap pMap;
+        if ( partition.canConvert( QVariant::Map ) )
+        {
+            pMap = partition.toMap();
+        }
+
+        // If it isn't a zfs partition, ignore it
+        if ( pMap[ "fsName" ] != "zfs" )
+        {
+            continue;
+        }
+
+        // Find the best device identifier, if one isn't available, skip this partition
+        QString deviceName;
+        if ( pMap[ "partuuid" ].toString() != "" )
+        {
+            deviceName = "/dev/disk/by-partuuid/" + pMap[ "partuuid" ].toString().toLower();
+        }
+        else if ( pMap[ "device" ].toString() != "" )
+        {
+            deviceName = pMap[ "device" ].toString().toLower();
+        }
+        else
+        {
+            continue;
+        }
+
+        // If the partition doesn't have a mountpoint, skip it
+        QString mountpoint = pMap[ "mountPoint" ].toString();
+        if ( mountpoint.isEmpty() )
+        {
+            continue;
+        }
+
+        // Build a poolname off config pool name and the mountpoint, this is not ideal but should work until there is UI built for zfs
+        QString poolName = m_poolName;
+        if ( mountpoint != '/' )
+        {
+            poolName += AlphaNumeric( mountpoint );
+        }
+
+        // Check to ensure the list of zfs info from the partition module is available and convert it to a list
+        if ( !gs->contains( "zfsInfo" ) && gs->value( "zfsInfo" ).canConvert( QVariant::List ) )
+        {
+            return Calamares::JobResult::error( tr( "Internal data missing" ), tr( "Failed to create zpool" ) );
+        }
+        QVariantList zfsInfoList = gs->value( "zfsInfo" ).toList();
```

Comment:
This chunk is done for every iteration of the loop, but isn't it supposed to be constant throughout the loop? If so, it can be lifted. That would also let you do the work, once, of collecting just the mount-points from the zfs info, so that the check below becoms a `QStringList.contains()` (or define a struct with mountpoint and bool encrypted, so you copy it from GS into a map / set / list just once).